waiter/src/waiter/state.clj

```diff
@@ -31,44 +31,86 @@
 
 ;; Router state maintainers
 
-(defn- healthy? [status-tags]
+(defn- healthy?
+  "Predicate on instances containing the :healthy status tag and optionally the :expired status tag."
+  [{:keys [status-tags]}]
   (or (= #{:healthy} status-tags)
       (= #{:healthy :expired} status-tags)))
 
+(defn expired?
+  "Predicate on instances containing the :expired status tag."
+  [{:keys [status-tags]}]
+  (contains? status-tags :expired))
+
 (defn- slots-available?
   "Predicate on healthy instances with available slots greater than used slots."
-  [{:keys [slots-assigned slots-used status-tags]}]
-  (and (healthy? status-tags) (> slots-assigned slots-used)))
+  [{:keys [slots-assigned slots-used] :as state}]
+  (and (healthy? state) (> slots-assigned slots-used)))
+
+(defn- only-lingering-requests?
+  "Returns truthy value if the instance is idle or processing expired requests based on the earliest-request-threshold-time"
+  [request-id->use-reason-map earliest-request-threshold-time]
+  (every? #(t/before? % earliest-request-threshold-time)
+          (->> request-id->use-reason-map
+               vals
+               (map :time))))
 
 (defn- killable?
-  "Only instances without slots used and not killed|locked can be killed."
-  [{:keys [slots-used status-tags]}]
-  (and (zero? slots-used)
-       (not-any? #(contains? status-tags %) [:killed :locked])))
+  "Only the following instances can be killed:
+   - without slots used and not killed|locked, or
+   - expired and idle or processing lingering requests.
+   If there are expired instances, only choose healthy instances that are blacklisted or expired.
+   If there are expired and starting instances, only kill unhealthy instances that are not starting."
+  [request-id->use-reason-map earliest-request-threshold-time expired-instances? starting-instances?
+   {:keys [slots-used status-tags] :as state}]
+  (and (or (and (zero? slots-used)
+                (not-any? #(contains? status-tags %) [:killed :locked]))
```

Comment:
Should we ever try killing a `:killed` or `:locked` instance?  Should this be a top-level condition?

Reply:
Moved to a top-level condition